mojin specific adjustments

lindemeier opened the pull request.

Commit:
minor enhancements

+ more warning and error messages indicating false configuration
save to bag given by sys arg or default to current dir


rosparam for duration of hardware settle

Commit:
finder library exported for public reuse

Commit:
use moving average to smooth measurements from checkerboard finder


always use averaged points in checkerboard detect

Commit:
support adding additional segments in KDL tree via yaml settings

Commit:
outputting calibration results as yaml and to std::out

renamed yaml output function

Commit:
improved logging

format fix

Commit:
clearing additional frames before insert

Commit:
create full dir tree for output

Commit:
output rotation as rpy instead of quaternion

makes more sense as offset angles

Commit:
added ros ok check in checkerboard finder loop

added more relevant checks and logging

fmessmer commented on robot_calibration/src/optimizer.cpp:
```diff
@@ -64,22 +66,71 @@ int Optimizer::optimize(OptimizationParams& params,
+  std::stringstream      segments_out;
+  const KDL::SegmentMap& segments = tree_.getSegments();
+  segments_out << "Segment name in kdl tree: " << std::endl;
+  for (KDL::SegmentMap::const_iterator it = segments.cbegin(); it != segments.cend(); it++)
```
`std::cbegin()` is `c++14`...this does not compile under `xenial`...
or you have to use `vector::cbegin`....i.e. add `#include <vector>`?

same with `std::cend()` 

see https://travis-ci.com/mojin-robotics/mojin_calibration/jobs/272377522

lindemeier replied:
Member function```std::vector::cbegin()``` is C++11 and the global function ```std::cbegin``` is C++14, yes. 
Wiil change it to C++98 mutable iterator ```std::vector::begin```.

Commit:
Removing const ref iterator (no C++11 or C++14 features)

Commit:
adds possibility to set the output folder

added observations message

added observation listener for generic observations

Revert "added observation listener for generic observations"

This reverts commit f1c2d5678e49f547927db70252b0b923a00d0eeb.

Revert "added observations message"

This reverts commit 838132fd9abec8316f56bd35f637a916e59eedc0.

removed image transport

Commit:
added circle grid support for checker board finder and more helper functions

tzpo fixed

Commit:
removing moving average filter

Commit:
bgr fixed

Revert "bgr fixed"

This reverts commit 37d0991c9fd99bb0b44af3417ad665ed09373c26.

fixed not found bug

Commit:
fixed circle pattern mix up of cols and rows

fixed compile error

Commit:
throw exception if chain not valid

Commit:
fixed flip in circle board object points generation

Commit:
Reverting capture_poses script to original state

lindemeier commented on robot_calibration/CMakeLists.txt:
```diff
@@ -59,13 +60,15 @@ catkin_package(
     sensor_msgs
     std_msgs
     tf
+    tf2_kdl
```
Why is this needed?

lindemeier replied:
```tf2::kdlToTransform``` is used in optimizer.cpp

lindemeier commented on robot_calibration/src/calibrate.cpp:
```diff
@@ -64,6 +65,117 @@
+    std::stringstream stream;
+
+    stream << "frames_offsets:" << std::endl;
+    for (uint32_t i = 0; i < params.free_frames.size(); i++)
```
unsigned vs signed

lindemeier commented on robot_calibration/src/calibrate.cpp:
```diff
@@ -64,6 +65,117 @@
+    }
+
+    stream << "offsets:" << std::endl;
+    for (uint32_t i = 0; i < params.free_params.size(); i++)
```
unsigned vs signed

lindemeier commented on robot_calibration/src/calibrate.cpp:
```diff
@@ -161,9 +281,7 @@ int main(int argc, char** argv)
-    for (unsigned pose_idx = 0;
-         (pose_idx < poses.size() || poses.size() == 0) && ros::ok();
-         ++pose_idx)
+    for (unsigned pose_idx = 0; (pose_idx < poses.size() || poses.size() == 0) && ros::ok(); ++pose_idx)
```
unsigned vs signed

lindemeier commented on robot_calibration/src/finders/checkerboard_finder.cpp:
```diff
@@ -109,10 +117,9 @@ bool CheckerboardFinder::waitForCloud()
 {
-  // Try up to 50 frames
-  for (int i = 0; i < 50; ++i)
+  for (int32_t i = 0U; (i < trials_) && (ros::ok()); ++i)
```
unsigned vs signed

fmessmer commented:
I'm happy to review...but I'd rather like you walk me through these changes face-to-face with some explanations...I don't know all the details...

lindemeier commented on robot_calibration/src/finders/checkerboard_finder.cpp:
```diff
@@ -197,40 +191,58 @@ bool CheckerboardFinder::findInternal(robot_calibration_msgs::CalibrationData *
     // Fill in message
     sensor_msgs::PointCloud2ConstIterator<float> xyz(cloud_, "x");
-    for (size_t i = 0; i < points.size(); ++i)
+    for (size_t i = 0; i < checker_board_image_positions.size(); ++i)
```
unsigned vs signed

lindemeier commented on robot_calibration/src/optimizer.cpp:
```diff
@@ -91,20 +142,41 @@ int Optimizer::optimize(OptimizationParams& params,
+    stream << "adding the following free params to the optimizer:" << std::endl;
+
+    // Setup  parameters to calibrate
+    for (size_t i = 0; i < params.free_params.size(); ++i)
```
unsigned vs signed

lindemeier replied:
done

lindemeier reviewed (commented):
unsigned vs signed issues

Commit:
signed vs unsigned assignment and comparison fixed

lindemeier commented:
> I'm happy to review...but I'd rather like you walk me through these changes face-to-face with some explanations...I don't know all the details...

Sure, would be better.

Commit:
output original fromat of calibration results to prevent breaking functionality

fmessmer committed:
nitpick review

fmessmer commented on robot_calibration/src/calibrate.cpp:
```diff
@@ -64,6 +65,126 @@
+    std::stringstream stream;
+
+    stream << "frames_offsets:" << std::endl;
+    for (uint32_t i = 0U; i < params.free_frames.size(); i++)
```
why don't you use `size_t` in for loops?
if you chose to, there is plenty of loops to apply this change...

lindemeier replied:
no specific reason for 32bit, only unsigned was intended. Changed it at locations where it makes sense to ```size_t```

fmessmer reviewed (dismissed):
formally looks ok :heavy_check_mark: 

I've updated `mojin-robotics/robot_calibration` `master` branch with the upstream changes from `mikeferguson/robot_calibration` and rebased your feature branch

please let me know whether you would like to change something about `size_t` (see below)
otherwise good to merge...
then it should be proposed to `mikeferguson` upstream repo

Commit:
use size_t rather than uint32_t for indices

fmessmer commented:
@lindemeier 
this is merged now
so you could:
- remove the overlay in `mojin_calibration`
- open a PR from `mojin-robotics` to `mikeferguson` (I think you should do it as you are the author of the additions)

lindemeier commented:
> @lindemeier
> this is merged now
> so you could:
> 
>     * remove the overlay in `mojin_calibration`
> 
>     * open a PR from `mojin-robotics` to `mikeferguson` (I think you should do it as you are the author of the additions)

- overlay removed
- PR will be created
